### PR TITLE
Add a remove-word command

### LINKED ISSUE #11: Adding downcased words to the dictionary and removing words
I think it will be nice to have the ability to send the downcased word to the dictionary. Aspell can do it with `&WORD` so maybe `&` can be used in addition to `@` to accept the word. This is maybe slightly inconsistent with `aspell` where the pairs is `*` and `&` but I don't think users need to know about aspell commands so it is fine.

It also makes sense to be able to remove a word from the personal dictionary since sometimes I do make further mistakes while correcting the previous ones. I don't see an aspell command for that in the info manual, but the format of the dictionary is really simple so this is just deleting a line and decrementing a number but I don't know if/how to reflect that deletion in the apelll process being used in `jit-spell`.

### PR DESCRIPTION
**Pull request by aikrahguzar**
Fix #11 

It took me some time due to copyright assignment and then a busy period but here is a version of the command that I think should work.

Some things to consider:

1) Right now it offers the current word and sub words as candidates. I think a better alternative would be to offer recent addition as candidates but they are not currently recorded but that is not too hard to change. 

2) Right now the command requires confirmation to select a candidate which I think is good since the syntax tables idea of current word might not agree with the spell-checker and this allows the user to modify the word to be removed. But maybe using prefix arg to directly use current word should be possible.

3)  When a word from personal dictionary is removed this basically issues a `jit-lock-refontify` in the buffer for every pending request. For the most likely scenario which at least for me is to correct a recent mistakenly added word this is overkill. Just starting a new process and copying the pending requests should be enough for that case. On the other hand if the word being removed is used in other buffers which have no pending request this would highlight those misspellings until recertification is done by some other means. If that scenario is a concern then the better way is go through the buffer list and issue `jit-lock-refontify` for every buffer which the process we are killing was being used. I am willing to go in either of those directions.

Thanks for the great package.

### CODE REVIEW

**Comment by astoff** [comment]
Thanks for PR. I hope I can have a look soon (I'm also somewhat busy at the moment).

**Comment by aikrahguzar** [comment]
> Thanks for PR. I hope I can have a look soon (I'm also somewhat busy at the moment).

No worries, please take your time.

File: jit-spell.el [hunk 1]
```diff
@@ -529,6 +529,50 @@ With a numeric ARG, move backwards that many misspellings."
...
+        (push "the buffer's word list" places)
+        (delete-region (match-beginning 1) (match-end 1))))
+    ;; Remove it from the local word list.
+    (when (member word jit-spell--local-words)
+      (cl-callf2 delete word jit-spell--local-words)
+      (unless places (push "the session's word list" places)))
+    ;; Remove it from the personal dictionary
+    (let ((case-fold-search nil))
+      (with-temp-buffer
+        (insert-file-contents ispell-personal-dictionary nil nil nil t)
```
Comment (astoff) [inline jit-spell.el:557]: I don't customize this variable, and apparently it's always nil to me. It seems you are assuming here that the variable is set to the spellchecker's default value in that case?

File: jit-spell.el [hunk 2]
```diff
@@ -529,6 +529,50 @@ With a numeric ARG, move backwards that many misspellings."
...
+            (kill-process proc)
+            (mapc (pcase-lambda (`(,buf _ _ _))
+                    (with-current-buffer buf (jit-lock-refontify)))
+                  pending)))))
+    (message "\"%s\" removed from %s." word (string-join places " and "))
+    (if (not places)
+        (message
+         "\"%s\" not found in the personal dictionary or local word lists."
+         word)
+      (jit-lock-refontify))))
```
Comment (astoff) [inline jit-spell.el:574]: Should this be as follows?

```suggestion
    (if (not places)
        (message
         "\"%s\" not found in the personal dictionary or local word lists."
         word)
      (message "\"%s\" removed from %s." word (string-join places " and "))
      (jit-lock-refontify))))
```

Comment (aikrahguzar) [reply jit-spell.el]: Thank you, I accepted the suggestion.

File: jit-spell.el (hunk 1 above)
Comment (aikrahguzar) [reply jit-spell.el:557]: That is a wrinkle, I had to set that variable for something long ago and forgot that it by default nil and more or less expected to be nil. I will try to figure out what is a reliable way of finding the default location and make changes accordingly.

Comment (aikrahguzar) [reply jit-spell.el:557]: Getting the location of a personal dictionary doesn't seem straightforward. See e.g. [this spell-fu issue](https://codeberg.org/ideasman42/emacs-spell-fu/issues/4).

For aspell the default location seems to be `~/.aspell.<lang>.pws`. I don't know where other programs store the personal word lists.

Comment (astoff) [reply jit-spell.el:557]: Okay, this depends too much on the spellchecker program and user configuration, so the only option is not to try to do anything about it.

Do you think the command is still meaningful even with this limitation? If so, please just add a test so that `insert-file-contents` doesn't fail if the var is nil / file doesn't exist.

(By the way, have you signed the FSF copyright assignment paperwork?)

Comment (aikrahguzar) [reply jit-spell.el]: > Okay, this depends too much on the spellchecker program and user configuration, so the only option is not to try to do anything about it.
> 
> Do you think the command is still meaningful even with this limitation? If so, please just add a test so that `insert-file-contents` doesn't fail if the var is nil / file doesn't exist.
> 

It is meaningful at least for me and I think setting finding the location of personal dictionary and setting `ispell-personal-dictionary` is doable. I added the limitation to the documentation though maybe we can also emit a warning once per session if the command is called without `ispell-personal-dictionary` being set.

Another limitation that I have realized is that the format of personal dictionaries is not standardized either. It seems like hunspell allows for affixes along with the words. For simple words all of `ispell`, `aspell` and `hunspell` all use one line per word. So the command will work for removing words that were added by jit-spell since `ispell -a` doesn't allow for affixes as far as I can tell.

> (By the way, have you signed the FSF copyright assignment paperwork?)

Yes, I have completed the assignment and Eli confirmed that it is on file a [while back](https://lists.gnu.org/archive/html/bug-gnu-emacs/2023-05/msg00751.html)

**Comment by astoff** [comment]
Is this the name and email that you want to appear in the version control history?

```
aikrahguzar <88336724+aikrahguzar@users.noreply.github.com>
```

**Comment by aikrahguzar** [comment]
> Thanks!
> 
> Is this the name and email that you want to appear in the version control history?
> 
> ```
> aikrahguzar <88336724+aikrahguzar@users.noreply.github.com>
> ```
> 
> Or should I rather use the info from the linked message to the emacs-devel list?

I am more or less not do not know the significance of this, I only started messing with git to contribute to emacs packages. But since `jit-spell` is on ELPA, and my copyright assignment is under `rahguzar@zohomail.eu`, it might be better to use that.

**Comment by astoff** [comment]
Okay, I now merged the command to the main branch. IMO having accurate author information on the commits is significant from the copyright perspective, and also to understand the history of the package in the future.
